apollo-router-core/src/services/checkpoint.rs

```diff
@@ -0,0 +1,299 @@
+    Response: Send + 'static,
+{
+    /// `CheckpointService` should
+    /// Forward the call to the next service
+    Forward(Request),
+    /// `CheckpointService` should not call the next service.
+    /// Return the provided Response instead
+    Return(Response),
+}
+
+#[allow(clippy::type_complexity)]
+pub struct CheckpointLayer<S, Request>
+where
+    S: Service<Request> + Send + 'static,
+    Request: Send + 'static,
+    S::Future: Send,
+    S::Response: Send + 'static,
+    S::Error: Send + 'static,
+{
+    checkpoint_fn: Arc<
+        dyn Fn(
+                Request,
+            ) -> Result<
+                Step<Request, <S as Service<Request>>::Response>,
+                <S as Service<Request>>::Error,
+            > + Send
+            + Sync
+            + 'static,
+    >,
+}
+
+#[allow(clippy::type_complexity)]
+impl<S, Request> CheckpointLayer<S, Request>
+where
+    S: Service<Request> + Send + 'static,
+    Request: Send + 'static,
+    S::Future: Send,
+    S::Response: Send + 'static,
+    <S as Service<Request>>::Error: Send + 'static,
+{
+    /// Create a `CheckpointLayer` from a function that takes a Service Request and returns a `Step`
+    pub fn new(
+        checkpoint_fn: impl Fn(
+                Request,
+            ) -> Result<
+                Step<Request, <S as Service<Request>>::Response>,
+                <S as Service<Request>>::Error,
+            > + Send
+            + Sync
+            + 'static,
+    ) -> Self {
+        Self {
+            checkpoint_fn: Arc::new(checkpoint_fn),
+        }
+    }
+}
+
+impl<S, Request> Layer<S> for CheckpointLayer<S, Request>
+where
+    S: Service<Request> + Send + 'static,
+    <S as Service<Request>>::Future: Send,
+    Request: Send + 'static,
+    <S as Service<Request>>::Response: Send + 'static,
+    <S as Service<Request>>::Error: Send + 'static,
+{
+    type Service = CheckpointService<S, Request>;
+
+    fn layer(&self, service: S) -> Self::Service {
+        CheckpointService {
+            checkpoint_fn: Arc::clone(&self.checkpoint_fn),
+            inner: service,
+        }
+    }
+}
+
+#[derive(Clone)]
+#[allow(clippy::type_complexity)]
+pub struct CheckpointService<S, Request>
+where
+    Request: Send + 'static,
+    S: Service<Request> + Send + 'static,
+    <S as Service<Request>>::Error: Send + 'static,
+    <S as Service<Request>>::Response: Send + 'static,
+    <S as Service<Request>>::Future: Send + 'static,
+{
+    inner: S,
+    checkpoint_fn: Arc<
+        dyn Fn(
+                Request,
+            ) -> Result<
+                Step<Request, <S as Service<Request>>::Response>,
+                <S as Service<Request>>::Error,
+            > + Send
+            + Sync
+            + 'static,
+    >,
+}
+
+#[allow(clippy::type_complexity)]
+impl<S, Request> CheckpointService<S, Request>
+where
+    Request: Send + 'static,
+    S: Service<Request> + Send + 'static,
+    <S as Service<Request>>::Error: Send + 'static,
```

Comment:
Can we follow the tower error types model?
All our layers should accept Into<BoxError> and output BoxError.

Reply:
fixed in ee986a9d

Reply:
fixed in 63b6a596